### PR TITLE
Implement span context and span builder; start cleaning up a bunch of other stuff

### PR DESCRIPTION
**Pull request by dazuma**

### CODE REVIEW

File: lib/opencensus/trace/span_builder.rb [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+module OpenCensus
+  module Trace
+    ##
+    # Span represents a single span within a request trace.
+    #
+    class SpanBuilder
+      ##
+      # The context that can build children of this span.
+      #
+      # @return [RequestTrace]
```
Comment (chingor13) [inline lib/opencensus/trace/span_builder.rb]: Is this a `SpanContext`?

File: lib/opencensus/trace/span_builder.rb [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      end
+
+      ##
+      # The numeric span ID of the parent, as a big-endian unsigned integer.
+      # May be 0 if this is a root span.
+      #
+      # @return [Integer]
+      #
+      def parent_span_id
+        context.parent.span_id
```
Comment (chingor13) [inline lib/opencensus/trace/span_builder.rb]: This will error if there's no parent

File: lib/opencensus/trace/span_context.rb [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        # unique `trace_id` and a root `span_id` of 0 is used.
+        #
+        # @param [String] header A Trace-Context header (optional)
+        # @param [Hash] rack_env The Rack environment hash (optional)
+        #
+        # @return [SpanContext]
+        #
+        def create_root header: nil, rack_env: nil
+          header ||= rack_env["HTTP_TRACE_CONTEXT"] if rack_env
+          trace_context = parse_trace_context_header header if header
```
Comment (chingor13) [inline lib/opencensus/trace/span_context.rb]: FYI, this format will be configurable as we'll need to support multiple formats

File: lib/opencensus/trace.rb [hunk 4]
```diff
@@ -16,42 +16,59 @@
...
       end
 
-      def spans
-        if parent = get
-          parent.spans
-        else
-          []
-        end
+      def start_span name
+        current_span_context.start_span name
```
Comment (chingor13) [inline lib/opencensus/trace.rb]: `current_span_context` could be nil

File: lib/opencensus/trace.rb [hunk 5]
```diff
@@ -16,42 +16,59 @@
...
       end
 
-      def set span
-        OpenCensus::Context.set CONTEXT_KEY, span
+      def in_span name, &block
+        current_span_context.in_span name, &block
```
Comment (chingor13) [inline lib/opencensus/trace.rb]: `current_span_context` could be nil

File: lib/opencensus/trace.rb [hunk 6]
```diff
@@ -16,42 +16,59 @@
...
       end
 
-      def get
-        OpenCensus::Context.get CONTEXT_KEY
+      def end_span
```
Comment (chingor13) [inline lib/opencensus/trace.rb]: Other implementations require providing the span we expect to restore as the current span. This mechanism is used to determine if we've gotten out of sync opening/closing spans.

File: lib/opencensus/trace/span_builder.rb (hunk 2 above)
Comment (dazuma) [reply lib/opencensus/trace/span_builder.rb]: There will always be a parent (the context that created this SpanBuilder is the parent).

File: lib/opencensus/trace/span_context.rb (hunk 3 above)
Comment (dazuma) [reply lib/opencensus/trace/span_context.rb]: Yes true. The struct is internal, and we'll deal with changes in the fields internally as needed.

File: lib/opencensus/trace.rb (hunk 4 above)
Comment (dazuma) [reply lib/opencensus/trace.rb]: Throws a more useful exception now.

File: lib/opencensus/trace.rb (hunk 5 above)
Comment (dazuma) [reply lib/opencensus/trace.rb]: Throws a more useful exception now.

File: lib/opencensus/trace.rb (hunk 6 above)
Comment (dazuma) [reply lib/opencensus/trace.rb]: Sounds good.

File: lib/opencensus/trace/span_builder.rb (hunk 1 above)
Comment (dazuma) [reply lib/opencensus/trace/span_builder.rb]: Whups, fixed.

**Pull request merged**